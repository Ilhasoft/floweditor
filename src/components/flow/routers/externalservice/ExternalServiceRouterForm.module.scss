@import 'variables.module.scss';

.link {
  cursor: pointer;
  color: $link_color;
  text-decoration: underline
}

.subject {
  margin-top: 8px;
}

.body {
  margin-top: 8px;
}