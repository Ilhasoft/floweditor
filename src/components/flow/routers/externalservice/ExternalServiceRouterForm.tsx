import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { nodeToState, stateToNode } from './helpers';
import { FormState, mergeForm, StringEntry, FormEntry } from 'store/nodeEditor';
import {
  Alphanumeric,
  Required,
  shouldRequireIf,
  StartIsNonNumeric,
  validate
} from 'store/validators';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import { Asset } from 'store/flowContext';
import * as React from 'react';
import i18n from 'config/i18n';
import TembaSelect from 'temba/TembaSelect';
import TypeList from 'components/nodeeditor/TypeList';
import { createResultNameInput } from 'components/flow/routers/widgets';
import ParamList from 'components/flow/routers/paramlist/ParamList';
import { ServiceCall } from 'config/interfaces';
import { ServicesCalls } from './constants';

export interface ExternalServiceRouterFormState extends FormState {
  externalService: FormEntry;
  call: FormEntry;
  resultName: StringEntry;
  calls: FormEntry;
  params: FormEntry;
}

export default class ExternalServiceRouterForm extends React.Component<
  RouterFormProps,
  ExternalServiceRouterFormState
> {
  constructor(props: RouterFormProps) {
    super(props);
    const externalServices = Object.values(this.props.assetStore.externalServices.items);
    const externalService = externalServices.length === 1 ? externalServices[0] : null;
    this.state = nodeToState(this.props.nodeSettings, externalService);

    bindCallbacks(this, {
      include: [/^handle/]
    });
  }
  private handleUpdate(
    keys: {
      externalService?: Asset;
      call?: ServiceCall;
      resultName?: string;
      calls?: any[];
      params?: any[];
    },
    submitting = false
  ): boolean {
    const updates: Partial<ExternalServiceRouterFormState> = {};

    if (keys.hasOwnProperty('externalService')) {
      updates.externalService = validate(
        i18n.t('forms.externalService', 'External Service'),
        keys.externalService,
        [shouldRequireIf(submitting)]
      );
    }

    if (keys.hasOwnProperty('call')) {
      updates.call = validate(i18n.t('forms.call', 'Call'), keys.call, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate(i18n.t('forms.result_name', 'Result Name'), keys.resultName, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('calls')) {
      updates.calls = validate(i18n.t('forms.calls', 'Calls'), keys.calls, [
        shouldRequireIf(submitting)
      ]);
    }

    // TODO: Improve paramter validation to check each field
    if (keys.hasOwnProperty('params')) {
      updates.params = validate(i18n.t('forms.params', 'Params'), keys.params, [
        shouldRequireIf(submitting)
      ]);
    }

    const updated = mergeForm(this.state, updates);

    this.setState(updated);
    return updated.valid;
  }

  private handleExternalServiceUpdate(selected: Asset[]): void {
    const newService = selected[0];
    this.handleUpdate({ externalService: newService, calls: ServicesCalls[newService.type] || [] });
  }

  private handleCallUpdate(call: ServiceCall): void {
    this.handleUpdate({ call });
  }

  private handleParamsUpdated(params: any[]): void {
    this.handleUpdate({ params });
  }

  private handleResultNameUpdate(value: string): void {
    const resultName = validate(i18n.t('forms.result_name', 'Result Name'), value, [
      Required,
      Alphanumeric,
      StartIsNonNumeric
    ]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  private handleSave(): void {
    // remove the last param as it is an empty one
    this.state.params.value.pop();

    const valid = this.handleUpdate(
      {
        externalService: this.state.externalService.value,
        call: this.state.call.value,
        params: this.state.params.value,
        resultName: this.state.resultName.value
      },
      true
    );

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  private renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const showExternalServices =
      Object.keys(this.props.assetStore.externalServices.items).length > 0 ||
      this.props.issues.length > 0;

    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        {showExternalServices ? (
          <div>
            <p>
              <span>Integration with... </span>
            </p>
            <AssetSelector
              key="select_external_service"
              name={i18n.t('forms.external_service', 'External Service')}
              placeholder="Select the external service to use"
              assets={this.props.assetStore.externalServices}
              onChange={this.handleExternalServiceUpdate}
              entry={this.state.externalService}
            />
          </div>
        ) : (
          ''
        )}
        <div>
          <p>
            <span>Action</span>
          </p>
          <TembaSelect
            key="select_external_service_call"
            name={i18n.t('forms.external_service_call', 'External Service Call')}
            placeholder="Select the call from external service to do"
            options={this.state.calls.value}
            nameKey="verboseName"
            onChange={this.handleCallUpdate}
            value={this.state.call.value}
            searchable={false}
          />
        </div>
        {this.state.call.value ? (
          <ParamList
            availableParams={this.state.call.value.params}
            params={this.state.params.value}
            onParamsUpdated={this.handleParamsUpdated}
          />
        ) : (
          <></>
        )}
        {createResultNameInput(this.state.resultName, this.handleResultNameUpdate)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
