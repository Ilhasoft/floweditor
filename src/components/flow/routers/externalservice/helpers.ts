import { createWebhookBasedNode } from 'components/flow/routers/helpers';
import { Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { CallExternalService } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, FormEntry } from 'store/nodeEditor';
import { createUUID } from 'utils';
import { ExternalServiceRouterFormState } from 'components/flow/routers/externalservice/ExternalServiceRouterForm';

export const getOriginalAction = (settings: NodeEditorSettings): CallExternalService => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 && settings.originalNode.node.actions[0]);

  if (action.type === Types.call_external_service) {
    return action as CallExternalService;
  }
};

export const nodeToState = (
  settings: NodeEditorSettings,
  initialExternalService: any
): ExternalServiceRouterFormState => {
  let externalService: FormEntry = initialExternalService
    ? { value: { uuid: initialExternalService.id, name: initialExternalService.name } }
    : { value: null };
  let body = { value: '' };
  let resultName = { value: 'Result' };
  let calls: any[] = initialExternalService.content.calls;
  let call = { value: calls[0] };

  if (getType(settings.originalNode) === Types.split_by_external_service) {
    const action = getOriginalAction(settings) as CallExternalService;
    externalService = { value: action.external_service };
    call = { value: action.call };
    body = { value: action.body };
    resultName = { value: action.result_name };
  }

  const state: ExternalServiceRouterFormState = {
    externalService,
    call,
    body,
    resultName,
    calls,
    valid: true
  };

  return state;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ExternalServiceRouterFormState
): RenderNode => {
  let uuid = createUUID();
  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const newAction: CallExternalService = {
    uuid,
    type: Types.call_external_service,
    external_service: {
      uuid: state.externalService.value.uuid,
      name: state.externalService.value.name
    },
    call: state.call.value,
    body: state.body.value,
    result_name: state.resultName.value
  };

  return createWebhookBasedNode(newAction, settings.originalNode, true);
};
