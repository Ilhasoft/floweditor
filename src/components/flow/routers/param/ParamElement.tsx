import { react as bindCallbacks } from 'auto-bind';
import FormElement from 'components/form/FormElement';
import TextInputElement, { TextInputStyle } from 'components/form/textinput/TextInputElement';
import { ServiceCallParam, ParamFilter } from 'config/interfaces';
import * as React from 'react';
import { FormState, StringEntry } from 'store/nodeEditor';

import styles from './ParamElement.module.scss';
import { initializeForm, validateParam } from './helpers';
import i18n from 'config/i18n';
import TembaSelect, { TembaSelectStyle } from 'temba/TembaSelect';
import { ParamProps } from '../paramlist/ParamList';

export interface ParamElementProps {
  initialParam: ParamProps;
  availableParams: ServiceCallParam[];
  name?: string; // satisfy form widget props
  onRemove?(uuid: string): void;
  onChange?(p: ParamProps): void;
}

export interface ParamElementState extends FormState {
  errors: string[];
  currentParam: ServiceCallParam;
  currentFilter: ParamFilter;
  paramFilters: ParamFilter[];
  data: StringEntry;
}

export default class ParamElement extends React.Component<ParamElementProps, ParamElementState> {
  constructor(props: ParamElementProps) {
    super(props);

    bindCallbacks(this, {
      include: [/^handle/]
    });

    this.state = initializeForm(props);
  }

  public componentDidMount() {
    const updates = validateParam({
      currentParam: this.state.currentParam,
      currentFilter: this.state.currentFilter,
      paramFilters: this.state.paramFilters,
      data: this.state.data
    });
    this.setState(updates as ParamElementState, () => this.handleChange());
  }

  private handleParamChange(newParam: ServiceCallParam): void {
    const updates = validateParam({
      currentParam: newParam,
      currentFilter: newParam.filters && newParam.filters.length >= 1 ? newParam.filters[0] : null,
      paramFilters: newParam.filters.filter(f => !f.required),
      data: this.state.data
    });
    this.setState(updates as ParamElementState, () => this.handleChange());
  }

  private handleFilterChange(newFilter: ParamFilter): void {
    const updates = validateParam({
      currentParam: this.state.currentParam,
      currentFilter: newFilter,
      paramFilters: this.state.paramFilters,
      data: this.state.data
    });
    this.setState(updates as ParamElementState, () => this.handleChange());
  }

  private handleDataChange(newData: string): void {
    const updates = validateParam({
      currentParam: this.state.currentParam,
      currentFilter: this.state.currentFilter,
      paramFilters: this.state.paramFilters,
      data: { value: newData }
    });
    this.setState(updates as ParamElementState, () => this.handleChange());
  }

  private handleRemoveClicked(): void {
    this.props.onRemove(this.props.initialParam.uuid);
  }

  private getParamProps(): ParamProps {
    const props: ParamProps = {
      uuid: this.props.initialParam.uuid,
      type: this.state.currentParam.type,
      verboseName: this.state.currentParam.verboseName,
      required: this.state.currentParam.required,
      filters: this.state.currentParam.filters,
      filter: { value: this.state.currentFilter },
      data: this.state.data,
      valid: this.state.valid
    };
    return props;
  }

  private handleChange(): void {
    this.props.onChange(this.getParamProps());
  }

  public render(): JSX.Element {
    const disableParam = this.state.currentParam && this.state.currentParam.required ? true : null;
    const disableFilter =
      this.state.currentFilter && this.state.currentFilter.required ? true : null;
    const canArrange = !disableParam && !disableFilter;

    const showFilter =
      this.state.currentParam &&
      this.state.currentParam.filters &&
      this.state.currentParam.filters.length;

    return (
      <FormElement
        data-spec="param-form"
        name={this.props.name}
        __className={styles.group}
        kaseError={this.state.errors.length > 0}
      >
        <div className={`${styles.param}`} data-draggable={canArrange}>
          {canArrange ? (
            <span className={`fe-chevrons-expand ${styles.dnd_icon}`} data-draggable={canArrange} />
          ) : (
            <div className={styles.order_filler}></div>
          )}
          <div className={styles.choice}>
            <TembaSelect
              name={i18n.t('forms.service_call_param', 'Service Call Param')}
              placeholder={i18n.t('forms.param', 'param')}
              style={TembaSelectStyle.small}
              options={this.props.availableParams}
              nameKey="verboseName"
              valueKey="name"
              disabled={disableParam || disableFilter}
              onChange={this.handleParamChange}
              value={this.state.currentParam}
            />
          </div>
          {showFilter ? (
            <div className={styles.choice}>
              <TembaSelect
                name={i18n.t('forms.service_call_param_filter', 'Service Call Param Filter')}
                placeholder={i18n.t('forms.filter', 'filter')}
                style={TembaSelectStyle.small}
                options={this.state.paramFilters}
                nameKey="verboseName"
                valueKey="name"
                onChange={this.handleFilterChange}
                value={this.state.currentFilter}
                disabled={disableFilter}
              />
            </div>
          ) : (
            <></>
          )}
          <div className={styles.data}>
            <TextInputElement
              name={i18n.t('forms.service_call_param_data', 'Service Call Param Data')}
              style={TextInputStyle.small}
              onChange={this.handleDataChange}
              entry={this.state.data}
            />
          </div>
          {!disableParam && !disableFilter ? (
            <span
              data-testid={'remove-param-' + this.props.initialParam.uuid}
              className={`fe-x ${styles.remove_icon}`}
              onClick={this.handleRemoveClicked}
            />
          ) : (
            <div className={styles.remove_filler}></div>
          )}
        </div>
      </FormElement>
    );
  }
}
