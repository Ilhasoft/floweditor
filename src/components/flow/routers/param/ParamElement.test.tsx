import ParamElement, { ParamElementProps } from 'components/flow/routers/param/ParamElement';
import * as React from 'react';
import { fireEvent, render, fireTembaSelect, fireChangeText } from 'test/utils';
import { createUUID } from 'utils';
import { ServicesCalls } from '../externalservice/constants';

const availableParams = ServicesCalls['omie'][0].params;

const paramUUID = createUUID();

const paramProps: ParamElementProps = {
  name: `param_${paramUUID}`,
  initialParam: {
    type: availableParams[0].type,
    verboseName: availableParams[0].verboseName,
    filters: availableParams[0].filters,
    uuid: paramUUID,
    filter: { value: availableParams[0].filters[0] },
    data: { value: '' },
    valid: true
  },
  availableParams,
  onRemove: jest.fn(),
  onChange: jest.fn()
};

describe(ParamElement.name, () => {
  describe('render', () => {
    it('should render param', () => {
      const { baseElement } = render(<ParamElement {...paramProps} />);
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('handles removes', () => {
      const onRemove = jest.fn();
      const { getByTestId } = render(<ParamElement {...paramProps} onRemove={onRemove} />);

      fireEvent.click(getByTestId(`remove-param-${paramProps.initialParam.uuid}`));
      expect(onRemove).toHaveBeenCalled();
    });

    it('handles data change', () => {
      const { baseElement, getByTestId } = render(
        <ParamElement {...paramProps} onRemove={jest.fn()} />
      );

      fireChangeText(getByTestId('Service Call Param Data'), 'new data');
      expect(baseElement).toMatchSnapshot();
    });

    it('handles filter change', () => {
      const { baseElement, getByTestId } = render(
        <ParamElement {...paramProps} onRemove={jest.fn()} />
      );

      fireTembaSelect(getByTestId('temba_select_service_call_param_filter'), [
        {
          name: 'cCodInt',
          type: 'text',
          verboseName: 'Código de Integração'
        }
      ]);
      expect(baseElement).toMatchSnapshot();
    });

    it('handles param change', () => {
      const { baseElement, getByTestId } = render(
        <ParamElement {...paramProps} onRemove={jest.fn()} />
      );

      fireTembaSelect(getByTestId('temba_select_service_call_param'), [availableParams[2]]);
      expect(baseElement).toMatchSnapshot();
    });

    it('handles filter change from correct param after param change', () => {
      const { baseElement, getByTestId } = render(
        <ParamElement {...paramProps} onRemove={jest.fn()} />
      );

      fireTembaSelect(getByTestId('temba_select_service_call_param'), [availableParams[2]]);
      fireTembaSelect(getByTestId('temba_select_service_call_param_filter'), [
        availableParams[2].filters[2]
      ]);
      fireChangeText(getByTestId('Service Call Param Data'), 'new data');
      expect(baseElement).toMatchSnapshot();
    });
  });
});
