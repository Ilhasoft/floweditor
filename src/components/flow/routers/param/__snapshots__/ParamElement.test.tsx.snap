// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParamElement render should render param 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="param"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param"
              errors="[]"
              expressions=""
              name="Service Call Param"
              namekey="verboseName"
              placeholder="param"
              valuekey="name"
              values="[{\\"type\\":\\"identificacao\\",\\"verboseName\\":\\"Identificação\\",\\"filters\\":[{\\"name\\":\\"nCod\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Contato\\"},{\\"name\\":\\"cCodInt\\",\\"type\\":\\"text\\",\\"verboseName\\":\\"Código de Integração\\",\\"required\\":true},{\\"name\\":\\"cNome\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Nome do contato\\",\\"required\\":true},{\\"name\\":\\"cSobrenome\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Sobrenome do contato\\"},{\\"name\\":\\"cCargo\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Cargo\\"},{\\"name\\":\\"dDtNasc\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Data de Nascimento\\"},{\\"name\\":\\"nCodVend\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Vendedor\\"},{\\"name\\":\\"nCodConta\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código da Conta\\"}],\\"uuid\\":\\"763e4844-3e1b-407a-a1b5-5fdfcd308b41\\",\\"filter\\":{\\"value\\":{\\"name\\":\\"nCod\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Contato\\"}},\\"data\\":{\\"value\\":\\"\\"},\\"valid\\":true}]"
            />
          </div>
        </div>
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param_filter"
              errors="[]"
              expressions=""
              name="Service Call Param Filter"
              namekey="verboseName"
              placeholder="filter"
              valuekey="name"
              values="[{\\"name\\":\\"nCod\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Contato\\"}]"
            />
          </div>
        </div>
        <div
          class="data"
        >
          <div
            data-use-vue-component-wrap=""
            style="all: unset;"
          >
            <div
              class="unnnic-text-area sm normal"
              data-passed-props="[object Object]"
              data-v-a95e6368=""
            >
              <!---->
              <textarea
                data-v-a95e6368=""
              />
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-param-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`ParamElement update handles data change 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="param"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param"
              errors="[]"
              expressions=""
              name="Service Call Param"
              namekey="verboseName"
              placeholder="param"
              valuekey="name"
              values="[{\\"type\\":\\"identificacao\\",\\"verboseName\\":\\"Identificação\\",\\"filters\\":[{\\"name\\":\\"nCod\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Contato\\"},{\\"name\\":\\"cCodInt\\",\\"type\\":\\"text\\",\\"verboseName\\":\\"Código de Integração\\",\\"required\\":true},{\\"name\\":\\"cNome\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Nome do contato\\",\\"required\\":true},{\\"name\\":\\"cSobrenome\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Sobrenome do contato\\"},{\\"name\\":\\"cCargo\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Cargo\\"},{\\"name\\":\\"dDtNasc\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Data de Nascimento\\"},{\\"name\\":\\"nCodVend\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Vendedor\\"},{\\"name\\":\\"nCodConta\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código da Conta\\"}],\\"uuid\\":\\"763e4844-3e1b-407a-a1b5-5fdfcd308b41\\",\\"filter\\":{\\"value\\":{\\"name\\":\\"nCod\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Contato\\"}},\\"data\\":{\\"value\\":\\"\\"},\\"valid\\":true}]"
            />
          </div>
        </div>
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param_filter"
              errors="[]"
              expressions=""
              name="Service Call Param Filter"
              namekey="verboseName"
              placeholder="filter"
              valuekey="name"
              values="[{\\"name\\":\\"nCod\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Contato\\"}]"
            />
          </div>
        </div>
        <div
          class="data"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="Service Call Param Data"
                name="Service Call Param Data"
                value="new data"
              />
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-param-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`ParamElement update handles filter change 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="param"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param"
              errors="[]"
              expressions=""
              name="Service Call Param"
              namekey="verboseName"
              placeholder="param"
              valuekey="name"
              values="[{\\"type\\":\\"identificacao\\",\\"verboseName\\":\\"Identificação\\",\\"filters\\":[{\\"name\\":\\"nCod\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Contato\\"},{\\"name\\":\\"cCodInt\\",\\"type\\":\\"text\\",\\"verboseName\\":\\"Código de Integração\\",\\"required\\":true},{\\"name\\":\\"cNome\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Nome do contato\\",\\"required\\":true},{\\"name\\":\\"cSobrenome\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Sobrenome do contato\\"},{\\"name\\":\\"cCargo\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Cargo\\"},{\\"name\\":\\"dDtNasc\\",\\"type\\":\\"string\\",\\"verboseName\\":\\"Data de Nascimento\\"},{\\"name\\":\\"nCodVend\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Vendedor\\"},{\\"name\\":\\"nCodConta\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código da Conta\\"}],\\"uuid\\":\\"763e4844-3e1b-407a-a1b5-5fdfcd308b41\\",\\"filter\\":{\\"value\\":{\\"name\\":\\"nCod\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código do Contato\\"}},\\"data\\":{\\"value\\":\\"\\"},\\"valid\\":true}]"
            />
          </div>
        </div>
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param_filter"
              errors="[]"
              expressions=""
              name="Service Call Param Filter"
              namekey="verboseName"
              placeholder="filter"
              valuekey="name"
              values="[{\\"name\\":\\"nCodConta\\",\\"type\\":\\"integer\\",\\"verboseName\\":\\"Código da Conta\\"}]"
            />
          </div>
        </div>
        <div
          class="data"
        >
          <div
            data-use-vue-component-wrap=""
            style="all: unset;"
          >
            <div
              class="unnnic-text-area sm normal"
              data-passed-props="[object Object]"
              data-v-a95e6368=""
            >
              <!---->
              <textarea
                data-v-a95e6368=""
              />
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-param-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`ParamElement update handles filter change from correct param after param change 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="param"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param"
              errors="[]"
              expressions=""
              name="Service Call Param"
              namekey="verboseName"
              placeholder="param"
              valuekey="name"
              values="[{\\"type\\":\\"telefone_email\\",\\"verboseName\\":\\"Telefone e Email\\",\\"filters\\":[{\\"name\\":\\"cDDDCel1\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do celular 1\\"},{\\"name\\":\\"cNumCel1\\",\\"type\\":\\"string\\",\\"maxLength\\":15,\\"verboseName\\":\\"Número do Celular 1\\"},{\\"name\\":\\"cDDDCel2\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do Celular 2\\"},{\\"name\\":\\"cNumCel2\\",\\"type\\":\\"string\\",\\"maxLength\\":15,\\"verboseName\\":\\"Número do Celular 2\\"},{\\"name\\":\\"cDDDTel\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do Telefone\\"},{\\"name\\":\\"cNumTel\\",\\"type\\":\\"string\\",\\"maxLength\\":15,\\"verboseName\\":\\"Número do Telefone\\"},{\\"name\\":\\"cDDDFax\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do Fax\\"},{\\"name\\":\\"cNumFax\\",\\"type\\":\\"string\\",\\"maxLength\\":15,\\"verboseName\\":\\"Número do Fax\\"},{\\"name\\":\\"cEmail\\",\\"type\\":\\"string\\",\\"maxLength\\":200,\\"verboseName\\":\\"E-mail do contato\\"},{\\"name\\":\\"cWebsite\\",\\"type\\":\\"string\\",\\"maxLength\\":100,\\"verboseName\\":\\"WebSite\\"}]}]"
            />
          </div>
        </div>
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param_filter"
              errors="[]"
              expressions=""
              name="Service Call Param Filter"
              namekey="verboseName"
              placeholder="filter"
              valuekey="name"
              values="[{\\"name\\":\\"cDDDCel2\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do Celular 2\\"}]"
            />
          </div>
        </div>
        <div
          class="data"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper small"
            >
              <input
                data-testid="Service Call Param Data"
                name="Service Call Param Data"
                value="new data"
              />
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-param-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`ParamElement update handles param change 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="param"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param"
              errors="[]"
              expressions=""
              name="Service Call Param"
              namekey="verboseName"
              placeholder="param"
              valuekey="name"
              values="[{\\"type\\":\\"telefone_email\\",\\"verboseName\\":\\"Telefone e Email\\",\\"filters\\":[{\\"name\\":\\"cDDDCel1\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do celular 1\\"},{\\"name\\":\\"cNumCel1\\",\\"type\\":\\"string\\",\\"maxLength\\":15,\\"verboseName\\":\\"Número do Celular 1\\"},{\\"name\\":\\"cDDDCel2\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do Celular 2\\"},{\\"name\\":\\"cNumCel2\\",\\"type\\":\\"string\\",\\"maxLength\\":15,\\"verboseName\\":\\"Número do Celular 2\\"},{\\"name\\":\\"cDDDTel\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do Telefone\\"},{\\"name\\":\\"cNumTel\\",\\"type\\":\\"string\\",\\"maxLength\\":15,\\"verboseName\\":\\"Número do Telefone\\"},{\\"name\\":\\"cDDDFax\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do Fax\\"},{\\"name\\":\\"cNumFax\\",\\"type\\":\\"string\\",\\"maxLength\\":15,\\"verboseName\\":\\"Número do Fax\\"},{\\"name\\":\\"cEmail\\",\\"type\\":\\"string\\",\\"maxLength\\":200,\\"verboseName\\":\\"E-mail do contato\\"},{\\"name\\":\\"cWebsite\\",\\"type\\":\\"string\\",\\"maxLength\\":100,\\"verboseName\\":\\"WebSite\\"}]}]"
            />
          </div>
        </div>
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_service_call_param_filter"
              errors="[]"
              expressions=""
              name="Service Call Param Filter"
              namekey="verboseName"
              placeholder="filter"
              valuekey="name"
              values="[{\\"name\\":\\"cDDDCel1\\",\\"type\\":\\"string\\",\\"maxLength\\":5,\\"verboseName\\":\\"DDD do celular 1\\"}]"
            />
          </div>
        </div>
        <div
          class="data"
        >
          <div
            data-use-vue-component-wrap=""
            style="all: unset;"
          >
            <div
              class="unnnic-text-area sm normal"
              data-passed-props="[object Object]"
              data-v-a95e6368=""
            >
              <!---->
              <textarea
                data-v-a95e6368=""
              />
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-param-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;
